### PR TITLE
Add the ability to search the website

### PR DESCRIPTION
**Pull request by logankilpatrick**
@tlienart when I try this out locally, I cannot get any search results to render but I am not sure what is wrong. Any suggestions?

### CODE REVIEW

**Comment by logankilpatrick** [comment]
Closes #64

File: website/_libs/lunr/lunrclient.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+// Parse search results into HTML
+function parseLunrResults(results) {
+    var html = [];
+    for (var i = 0; i < results.length; i++) {
+        var id = results[i]["ref"];
+        var item = PREVIEW_LOOKUP[id]
+        var title = item["t"];
+        var preview = item["p"];
+        var link = item["l"].replace("__site/", "");
```
Comment (logankilpatrick) [inline website/_libs/lunr/lunrclient.js:32]: would this need to change?

**Comment by tlienart** [comment]
This is almost good to go, just in your `head.html` you need to include the other JS files, to make it simpler, copy paste:

```html
  <!-- LUNR -->
  <script src="/libs/lunr/lunr.min.js"></script>
  <script src="/libs/lunr/lunr_index.js"></script>
  <script src="/libs/lunr/lunrclient.min.js"></script>
```

You might want to edit the `deploy` script to include a `lunr` step as well so that whenever the site gets updated the index does too; for this you'd need to replace this line https://github.com/mitmath/18S191/blob/e4dd66ebb41e442202633881a7f626aa6525dee1/.github/workflows/ExportNotebooks.yml#L55 by something like

```
run(`$(npm_cmd()) install lunr`);
run(`$(npm_cmd()) install cheerio`);
serve(single=true, cleanup=false); lunr(); optimize(minify=false);
```

File: website/_libs/lunr/lunrclient.js (hunk 1 above)
Comment (tlienart) [reply website/_libs/lunr/lunrclient.js:32]: no, this is fine :+1:

File: website/_layout/head.html [hunk 2]
```diff
@@ -7,6 +7,10 @@
   {{if hascode}} {{insert head_highlight.html }}   {{end}}
   {{insert style.html}}
   {{isdef title}} <title>{{fill title}}</title>  {{end}}
+
+  <!-- LUNR -->
+  <script src="https://unpkg.com/lunr/lunr.js"></script>
```
Comment (tlienart) [inline website/_layout/head.html:12]: replace this by the block of 4 script lines, to include not only the client but also the index!

**Comment by logankilpatrick** [comment]
Added those, however, it looks like things still don't work as expected, this is what I get: 

Screen Shot 2021-08-24 at 6 59 26 AM

And the actual links 404

**Comment by tlienart** [comment]
this is on local side? (I've got this working on a local branch)

**Comment by logankilpatrick** [comment]
Yes, on the local side that is what the preview looks like

**Comment by tlienart** [comment]
so on my side I also get that all links show as `Introduction to computational thinking` (will check how you can fix this) but the links don't error. 

Just to make sure we're at the same point, can you search for "addition" and confirm there's two links? (locally)

**Comment by logankilpatrick** [comment]
Yes, I get two links total, but when I click one, it says: "404: file not found. Perhaps you made a typo in the URL, or the requested file has been deleted or renamed." and the URL is "http://localhost:8000/website/logistics/index.html".

**Comment by logankilpatrick** [comment]
I am running with regular `serve()`, could that be the issue?

**Comment by tlienart** [comment]
no no this is fine; ok let me dig a bit

**Comment by tlienart** [comment]
ok so the local fail is due to `prepath`, that's fine and is a known issue (locally for now you can comment out `prepath` in the config.md and you'll see that stuff works)

I'm looking into the link title thing

**Comment by logankilpatrick** [comment]
What additional stuff should work? It seems like the same results to me

**Comment by tlienart** [comment]
Ok so what you need to fix is going to be a little bit annoying, you need to remove 

```
@def title = ...
```

in the config.md, and add a `@def title` to every page that doesn't have a first `#` or already an explicitly specified `#`. so in particular all `hw*.md` will need a `@def title = ...` 

If you do this then things will work fine (don't forget to eventually uncomment the `prepath = ...` in the `config.md` when you push here)

Screenshot 2021-08-24 at 17 00 38

**Comment by tlienart** [comment]
> What additional stuff should work? It seems like the same results to me

you need to rebuild the index so, locally, after having commented out the prepath, 

```
julia> serve(single=true, cleanup=false); lunr(); serve()
```

**Comment by logankilpatrick** [comment]
Okay cool. should be easy enough! Thanks for figuring this out, I will make these updates in the next few days and make sure all is good before pushing.

**Comment by tlienart** [comment]
To make sure I'm completely clear here's what I'm suggesting

## Locally

* comment out `prepath` in config.md 
* comment out `title` in config.md
* add a `title` def in all `.md` pages except those that already have a top section `#` for which it's automatic (you can also specify a `title` for those, it will take precedence over the first top section); in particular this is needed for all `hw*.md` 
* rebuild, serve and test: `serve(single=true, cleanup=false); lunr(); serve()`

If that looks fine then

* uncomment `prepath` in config.md
* add the following the `.gitignore`  (node stuff)

```
node_modules/
package-lock.json
```

* make sure that the same line `serve(single=true, cleanup=false); lunr(); optimize(minify=false)`  is in the deploy script
* add, commit, push and merge

**Comment by fonsp** [comment]
Is this able to search the contents of notebooks? That would be really cool! If not, I'd like to take a look at it sometime

**Comment by tlienart** [comment]
I don't see why it couldn't, maybe you have to tell it to also process .ipynb files but otherwise I imagine it's doable (also maybe of interest: https://blog.ouseful.info/2018/05/10/initial-sketch-searching-jupyter-notebooks-using-lunr/ ?)

**Comment by fonsp** [comment]
We use Pluto notebooks :)

**Comment by tlienart** [comment]
hahaha of course, I think the comment holds though, Lunr should be able to index any readable file afaik

**Comment by fonsp** [comment]
Superseded by #124

**Pull request closed without merging**